tsserver loads files as inferred projects, not the real tsconfig project

find_referencing_symbols returns 0 for TypeScript while find_symbol/definition work 

Preconditions:

`find_referencing_symbols` / `request_references` returns **0 references** for TypeScript symbols that have many real cross-file callers, while `find_symbol`, `get_symbols_overview`, and `request_definition` all work correctly on the same running server. Python references (pyright) work fine in the same project.

I traced the LSP wire protocol: the references request is well-formed and points at the correct identifier position, tsserver does ~2s of real work, then returns `count=0`. This appears to be the unresolved TS tail of #478. Diagnosis: **tsserver is treating each `didOpen`'d file as a standalone inferred project and never loads the package's `tsconfig.json`**, so find-references has no project graph to search.
Issue details:

## Setup
- Serena: **1.5.3** (latest on PyPI)
- MCP client: Claude Code
- typescript-language-server: **5.1.3**; typescript/tsserver: **5.9.3**
- Node: **v22.20.0**; OS: Linux x86_64 (kernel 6.1)
- Language backend: `typescript` (also reproduced with `typescript_vts`)
- Project: TS monorepo, **no root `tsconfig.json`** — each package owns its own. The target package's `tsconfig.json` is plain `include`-based (`"include": ["src/**/*.ts", ...]`), **no `composite`**, with one project `references` entry to a sibling package. ~21k files total.
## Reproduction

Target: an exported `const fromNumber` in `<pkg>/src/util/mathCore.ts`, with 76 in-package caller files, all under `<pkg>/src/`, imported via plain relative paths (`import { fromNumber } from '../util/mathCore'`).

1. Create a `SolidLanguageServer` (mirroring `ls_manager.create_language_server`) with `repository_root_path` = the package root (so its tsconfig is THE project) and `additional_workspace_folders=[]`.
2. `request_document_symbols(file)` → locate the symbol (works; returns it at line 42, char 13).
3. `request_references(file, line=42, column=13)` → returns `[]`.

## Wire trace (the key evidence)

Request is correct; tsserver returns count=0 after real work:

```
solidlsp -> ls: {"jsonrpc":"2.0","method":"textDocument/references","id":4,
  "params":{"textDocument":{"uri":"file://.../<pkg>/src/util/mathCore.ts"},
  "position":{"line":42,"character":13},"context":{"includeDeclaration":false}}}
solidlsp.ls perf: request_references path=src/util/mathCore.ts elapsed_ms=2029 count=0
```

On the same server: `request_definition` returns a hit, and `request_references` for a **Python** symbol returns correct cross-file callers — so the client plumbing is fine; this is specific to the TS server's find-references.

## What I tried (all still return 0)

- `additional_workspace_folders` set to the package roots
- Package opened **as the LSP root** (its tsconfig primary)
- Pre-opening **all 75 caller files** via `request_document_symbols` + a 30s wait
- `typescript_vts` (vtsls) backend
- `serena project index`
- Raising `INDEXING_PROGRESS_TIMEOUT`
- Confirmed neither the symbol nor its callers sit under an ignored dir name (`node_modules`/`dist`/`build`/`coverage`, re: #1523)

## Suspected cause + questions

tsserver only returns references for files in a **loaded project**. When solidlsp `didOpen`s a `.ts` file it seems to land in an **inferred (single-file) project** instead of tsserver loading the owning `tsconfig.json`, so find-references has nothing to search. Definition still works because it follows imports forward.

1. Does solidlsp signal tsserver to **load the tsconfig project** for an opened file (e.g. `_typescript.configurePlugin`, `openExternalProject`, or relying on upward `tsconfig.json` auto-detection)? On this layout it appears to stay inferred.
2. Is find-references known to require `composite`/project-references or a specific `initializationOptions`? Ours is a plain `include`-based config.
3. Your `test_typescript_cross_package` fixture passes — how does it differ structurally (composite? root tsconfig?) from a no-root-tsconfig, `include`-based package?

It shouldn't require having to open all files, and a bug this serious would have been caught by the many users of our TS-server-based integrations, I think. 

There might be something flaky or a race condition. In CI we occasionally get

```
FAILED test/serena/test_serena_agent.py::TestSerenaAgent::test_find_symbol_references[typescript_helper_refs] - AssertionError: Expected to find reference to helperFunction in use_helper.ts. refs={'index.ts': {'File': [{'name_path': 'index', 'body_location': {'start_line': 0, 'end_line': 26}, 'content_around_reference': '...  19:\n  >  20:helperFunction();\n...  21:'}]}}
assert False
 +  where False = <function TestSerenaAgent.test_find_symbol_references.<locals>.contains_ref_with_relative_path at 0x109f2eb60>({'index.ts': {'File': [{'body_location': {'end_line': 26, 'start_line': 0}, 'content_around_reference': '...  19:\n  >  20:helperFunction();\n...  21:', 'name_path': 'index'}]}}, 'use_helper.ts')
 +    where 'use_helper.ts' = FindReferenceCase(language=<Language.TYPESCRIPT: 'typescript'>, id='typescript_helper_refs', symbol_name='helperFunction', definition_file='index.ts', reference_file='use_helper.ts').reference_file
```

See here for example
https://github.com/oraios/serena/actions/runs/28047269341/job/83028449968

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,15 @@ Status of the `main` branch. Changes prior to the next official version change w
 * JetBrains:
   - `jet_brains_find_symbol`: Disallow wildcard-only search, delegating to overview tool if request is for file
 
+* Language Servers:
+  - `typescript`: Fix: on large projects, the first `find_referencing_symbols`/`request_references` call
+    could silently race tsserver's project load and return incomplete results, because the fixed 2s
+    grace for tsserver to *start* reporting `$/progress` (distinct from the separate, already
+    configurable `indexing_timeout` used to wait for it to *drain*) was hardcoded and not large enough
+    for projects where the initial project-graph resolution itself takes longer than that. The grace
+    is now `indexing_start_grace` (default 5.0s), configurable the same way as `indexing_timeout` and
+    `server_ready_timeout` #1586
+
 * Hooks:
   - Add `serena-hooks --client=grok`, including Grok-native PreToolUse allow/deny output.
   - PreToolUse remind hook: coerce non-string shell command values instead of failing, and recognize

--- a/docs/02-usage/050_configuration.md
+++ b/docs/02-usage/050_configuration.md
@@ -1118,8 +1118,9 @@ Supported settings:
 | `typescript_version` | `5.9.3` | Override the bundled `typescript` npm package version Serena installs when `ls_path` is not set. |
 | `typescript_language_server_version` | `5.1.3` | Override the bundled `typescript-language-server` npm package version Serena installs when `ls_path` is not set. |
 | `npm_registry` | `null` | Override the npm registry Serena uses for the managed install. |
-| `indexing_timeout` | `30.0` | Timeout in seconds for waiting on tsserver's `$/progress` project-indexing signal (both at startup and before the first cross-file reference query). If indexing does not complete within this window, Serena logs a warning and proceeds anyway. Increase it for very large projects. |
+| `indexing_timeout` | `30.0` | Timeout in seconds for waiting on tsserver's `$/progress` project-indexing signal to *drain* once it has started (both at startup and before the first cross-file reference query). If indexing does not complete within this window, Serena logs a warning and proceeds anyway. Increase it for very large projects. |
 | `server_ready_timeout` | `10.0` | Timeout in seconds for waiting on the server-ready signal after initialization. If the signal does not arrive within this window, Serena logs a message and proceeds anyway. |
+| `indexing_start_grace` | `5.0` | Timeout in seconds to wait for tsserver to *start* reporting `$/progress` before the first cross-file reference query. tsserver must resolve the project graph before it can emit the first progress token, and that can take longer than the default on a very large project; if it takes longer than this window, Serena assumes no indexing was needed and may return incomplete cross-file references. Raising `indexing_timeout` alone does not help here, since this grace elapses first. Increase this for very large projects if `find_referencing_symbols`/`request_references` returns incomplete results shortly after project load. |
 
 #### Svelte
 

--- a/src/solidlsp/language_servers/typescript_language_server.py
+++ b/src/solidlsp/language_servers/typescript_language_server.py
@@ -74,6 +74,8 @@ class TypeScriptLanguageServer(SolidLanguageServer):
         - typescript_language_server_version: Version of typescript-language-server to install (default: "5.1.3")
         - indexing_timeout: float, timeout in seconds for project indexing (default: 30.0)
         - server_ready_timeout: float, timeout in seconds for the server-ready signal (default: 10.0)
+        - indexing_start_grace: float, timeout in seconds to wait for tsserver to *start*
+          reporting indexing progress before the first cross-file reference query (default: 5.0)
     """
 
     @classmethod
@@ -84,6 +86,11 @@ class TypeScriptLanguageServer(SolidLanguageServer):
     # well within this window; the timeout is only hit if the server never sends progress.
     INDEXING_PROGRESS_TIMEOUT = 30.0
     SERVER_READY_TIMEOUT = 10.0
+    # How long to wait for tsserver to *start* emitting $/progress after opening files, before
+    # assuming no indexing is needed. tsserver has to resolve the project graph before it can even
+    # create the first progress token, and that resolution is proportional to project size, so a
+    # large project can genuinely take longer than a small one just to begin reporting.
+    INDEXING_START_GRACE = 5.0
 
     def __init__(self, config: LanguageServerConfig, repository_root_path: str, solidlsp_settings: SolidLSPSettings):
         """
@@ -123,7 +130,7 @@ class TypeScriptLanguageServer(SolidLanguageServer):
         :param start_grace: Maximum seconds to wait for progress to begin after opening files.
         :return: True if indexing completed or no progress began within the grace period, False on timeout.
         """
-        grace = self._INDEXING_START_GRACE_S if start_grace is None else start_grace
+        grace = self.INDEXING_START_GRACE if start_grace is None else start_grace
 
         # wait for progress to begin
         progress_deadline = time.monotonic() + grace
@@ -170,6 +177,10 @@ class TypeScriptLanguageServer(SolidLanguageServer):
     def _get_indexing_timeout(self) -> float:
         """:return: maximum seconds to wait for TypeScript project indexing."""
         return float(self._custom_settings.get("indexing_timeout", self.INDEXING_PROGRESS_TIMEOUT))
+
+    def _get_indexing_start_grace(self) -> float:
+        """:return: maximum seconds to wait for tsserver to start reporting indexing progress."""
+        return float(self._custom_settings.get("indexing_start_grace", self.INDEXING_START_GRACE))
 
     def _handle_server_ready_timeout(self, timeout: float) -> None:
         """Handle a TypeScript server-ready timeout.
@@ -531,17 +542,14 @@ class TypeScriptLanguageServer(SolidLanguageServer):
         if not self._has_waited_for_cross_file_references:
             self.expect_indexing()
 
-    #: how long to wait for tsserver to *start* reporting indexing progress after didOpen;
-    #: matches the fixed wait the base implementation previously used
-    _INDEXING_START_GRACE_S = 2.0
-
     @override
     def _wait_for_cross_file_references_if_needed(self) -> None:
         if self._has_waited_for_cross_file_references:
             return
 
         timeout = self._get_indexing_timeout()
-        if self._wait_for_indexing_start_or_completion(timeout=timeout):
+        start_grace = self._get_indexing_start_grace()
+        if self._wait_for_indexing_start_or_completion(timeout=timeout, start_grace=start_grace):
             log.info("TypeScript cross-file indexing complete")
         else:
             log.warning(

--- a/test/solidlsp/test_typescript_timeout_policy.py
+++ b/test/solidlsp/test_typescript_timeout_policy.py
@@ -63,6 +63,14 @@ class TestBaseTypeScriptTimeoutPolicy:
         assert server._get_server_ready_timeout() == 1.5
         assert server._get_indexing_timeout() == 2.5
 
+    def test_indexing_start_grace_default(self) -> None:
+        server = _bare_ts_server(TypeScriptLanguageServer)
+        assert server._get_indexing_start_grace() == 5.0
+
+    def test_indexing_start_grace_configurable_via_ls_specific_settings(self) -> None:
+        server = _bare_ts_server(TypeScriptLanguageServer, {"indexing_start_grace": 12.0})
+        assert server._get_indexing_start_grace() == 12.0
+
 
 class TestSvelteCompanionTimeoutPolicy:
     """The Svelte companion must be strict: raise instead of serving from a cold/partial program."""
@@ -139,6 +147,36 @@ class TestWaitForIndexingStartOrCompletion:
             assert server._wait_for_indexing_start_or_completion(timeout=30.0, start_grace=0.0) is True
         finally:
             timer.cancel()
+
+
+class TestWaitForCrossFileReferencesUsesConfiguredGrace:
+    """The actual find-references call path (not just the helper in isolation) must honor
+    indexing_start_grace: this is the mechanism behind oraios/serena#1586, where a large
+    project's tsserver takes longer than the (previously hardcoded, unconfigurable) grace to
+    even start reporting progress, so the first cross-file reference query silently races a
+    still-loading project and returns incomplete results.
+    """
+
+    def test_configured_grace_bounds_the_real_call_path(self) -> None:
+        server = _bare_ts_server(TypeScriptLanguageServer, {"indexing_start_grace": 0.05})
+        server._has_waited_for_cross_file_references = False
+        server.expect_indexing()  # mirrors _pre_open_for_cross_file_references having already run
+
+        start = time.monotonic()
+        server._wait_for_cross_file_references_if_needed()
+        elapsed = time.monotonic() - start
+
+        assert server._has_waited_for_cross_file_references is True
+        # bounded by the configured 0.05s grace, not the 5.0s default (let alone the old 2.0s one)
+        assert elapsed < 1.0
+
+    def test_second_call_is_a_noop_regardless_of_grace(self) -> None:
+        server = _bare_ts_server(TypeScriptLanguageServer, {"indexing_start_grace": 0.05})
+        server._has_waited_for_cross_file_references = True
+
+        start = time.monotonic()
+        server._wait_for_cross_file_references_if_needed()
+        assert time.monotonic() - start < 0.1
 
 
 class _FakeSolidLSPSettings:

